iothub/service/src/Amqp/AmqpClientHelper.cs

```diff
@@ -33,109 +31,170 @@ public static Exception ToIotHubClientContract(Exception exception)
+        public static Exception GetExceptionFromOutcome(Outcome outcome)
         {
-            Logging.Enter(faultTolerantReceivingLink, deliveryTag, outcome.DescriptorCode, batchable, nameof(DisposeMessageAsync));
+            if (outcome == null)
+            {
+                return new IotHubException("Unknown error.");
+            }
 
-            try
+            Exception retException;
+            if (outcome.DescriptorCode == Rejected.Code)
+            {
+                var rejected = (Rejected)outcome;
+                retException = ToIotHubClientContract(rejected.Error);
+            }
+            else if (outcome.DescriptorCode == Released.Code)
+            {
+                retException = new OperationCanceledException("AMQP link released.");
+            }
+            else
             {
-                cancellationToken.ThrowIfCancellationRequested();
+                retException = new IotHubException("Unknown error.");
+            }
 
-                Outcome disposeOutcome;
-                try
-                {
-                    ReceivingAmqpLink deviceBoundReceivingLink = await faultTolerantReceivingLink.GetReceivingLinkAsync().ConfigureAwait(false);
-                    disposeOutcome = await deviceBoundReceivingLink
-                        .DisposeMessageAsync(
-                            deliveryTag,
-                            outcome,
-                            batchable,
-                            cancellationToken)
-                        .ConfigureAwait(false);
-                }
-                catch (Exception exception)
-                {
-                    Logging.Error(faultTolerantReceivingLink, exception, nameof(DisposeMessageAsync));
+            return retException;
+        }
 
-                    if (exception.IsFatal())
-                    {
-                        throw;
-                    }
+        public static Exception ToIotHubClientContract(Error error)
+        {
+            if (error == null)
+            {
+                return new IotHubException("Unknown error.");
+            }
 
-                    throw ToIotHubClientContract(exception);
-                }
+            Exception retException;
+            string message = error.Description;
+            string trackingId = null;
 
-                Logging.Info(faultTolerantReceivingLink, disposeOutcome.DescriptorCode, nameof(DisposeMessageAsync));
+            if (error.Info != null
+                && error.Info.TryGetValue(AmqpsConstants.TrackingId, out trackingId))
+            {
+                message = $"{message}\r\nTracking Id:{trackingId}";
+            }
 
-                if (disposeOutcome.DescriptorCode != Accepted.Code)
-                {
-                    throw AmqpErrorMapper.GetExceptionFromOutcome(disposeOutcome);
-                }
+            if (error.Condition.Equals(IotHubAmqpErrorCode.TimeoutError))
+            {
+                retException = new TimeoutException(message);
```

Comment:
I'm hoping we can find a way to not throw TimeoutException any longer. I was talking to Bryce about this today.

Any insights? When does it happen? I'm guessing this isn't the same as OperationCanceledException?

Reply:
This is copied verbatim from our current error mapping logic to a new file. I'd like to save this particular discussion for the Exception PRs that Bryce will be driving